test/e2e/storage/volume_metrics.go

```diff
@@ -143,15 +146,20 @@ var _ = utils.SIGDescribe("[Serial] Volume metrics", func() {
 		err = e2epod.WaitTimeoutForPodRunningInNamespace(c, pod.Name, pod.Namespace, f.Timeouts.PodStart)
 		framework.ExpectNoError(err, "Error starting pod %s", pod.Name)
 
-		framework.Logf("Deleting pod %q/%q", pod.Namespace, pod.Name)
 		framework.ExpectNoError(e2epod.DeletePodWithWait(c, pod))
 
 		updatedStorageMetrics := waitForDetachAndGrabMetrics(storageOpMetrics, metricsGrabber, pluginName)
 
 		framework.ExpectNotEqual(len(updatedStorageMetrics.latencyMetrics), 0, "Error fetching c-m updated storage metrics")
 		framework.ExpectNotEqual(len(updatedStorageMetrics.statusMetrics), 0, "Error fetching c-m updated storage metrics")
 
-		volumeOperations := []string{"volume_provision", "volume_detach", "volume_attach"}
+		volumeOperations := []string{"volume_detach", "volume_attach"}
+		// volume_provision is only emitted from an intree plugin,
+		// CSI Plugins go through a different path that doesn't emit this metric
+		// see pkg/controller/volume/persistentvolume/pv_controller.go:provisionClaim
+		if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CSIMigration) {
```

Comment:
feature gate checking does not work in e2e tests because it doesn't actually check what is enabled on the server.

Have you tried running this test with csi migration and you see it failing for the provision metric?

Reply:
I didn't and I'm going to try that next, I'm writing this comment for reference:

I created my cluster with:

```
cat /etc/default/kubelet
KUBELET_OPTS="--v=4 --feature-gates=KubeletPodResources=false,CSIMigrationGCE=false
```

And when I run the test before adding this check the e2e test was checking for `volume_provision` which is emitted in case the volume is provisioned with an intree plugin, according to these lines

https://github.com/kubernetes/kubernetes/blob/a5622f3f6e1f5d5ffd066d799204ea0c6d1a49e2/pkg/controller/volume/persistentvolume/pv_controller.go#L1455-L1475

So because `CSIMigrationGCE=false` it should be provisioned with the intree plugin, however I think as you say my config might not be valid, e.g. CSIMigrationGCE=false, PD CSI Driver installed with a default StorageClass that's pointing to it. I think that what happened in 

https://github.com/kubernetes/kubernetes/blob/a5622f3f6e1f5d5ffd066d799204ea0c6d1a49e2/pkg/controller/volume/persistentvolume/pv_controller.go#L1803-L1814

is that `plugin, err := ctrl.volumePluginMgr.FindProvisionablePluginByName(class.Provisioner)` didn't return the plugin because the StorageClass that I installed points to the external PD CSI Driver storage class (the one that comes from the PD CSI repo and not from `./cluster/addons`) so it correctly went through the external provision path even with CSIMigrationGCE=false!

Reply:
also for reference, I took `utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CSIMigration)` from https://github.com/kubernetes/kubernetes/blob/a5622f3f6e1f5d5ffd066d799204ea0c6d1a49e2/test/e2e/storage/volume_limits.go#L39

Reply:
created a cluster with CSIMigrationGCE set to true:

```
	CSIMigrationGCE:                                {Default: true, PreRelease: featuregate.Beta}, // On by default in 1.23 (requires GCE PD CSI Driver)
```

Removed this `if` so that the metrics checked include `volume_provision`, recompiled and run the e2e tests and they failed.

Next, I kept these guards, recompiled and run the tests, the test passed.

It seems like `		if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CSIMigration) {
` works

`utilfeature.DefaultMutableFeatureGate` seems to be just an objects that also exists when the test binary runs and from https://github.com/kubernetes/kubernetes/blob/a5622f3f6e1f5d5ffd066d799204ea0c6d1a49e2/cmd/kubelet/app/server.go#L186

looks like it's populated with the kubelet config

Reply:
my bad, I think I'm mixing the kubelet feature gate object and the e2e test feature object, I'll keep on checking why it works

Reply:
You need to set feature gates on all kubernetes components, not just kubelet.

This feature gate function is just a library. It only indicates if the feature gate is enabled in the binary it's built in. It cannot check if a feature is actually enabled in a cluster.

Reply:
ah, it works because the default value of CSIMigrationGCE is true regardless of where it's created e.g. kubelet or e2e, lol

Reply:
If you are using kubetest to bring up a gce cluster, there should be a flag/env to enable feature gates.

Reply:
> This feature gate function is just a library. It only indicates if the feature gate is enabled in the binary it's built in. It cannot check if a feature is actually enabled in a cluster.

yep I got it

> If you are using kubetest to bring up a gce cluster, there should be a flag/env to enable feature gates.

for this test luckily we don't need to do any feature gate setup as the defaults now reflect the current state e.g. CSIMigration=true, CSIMigrationGCE=true

Reply:
What happens if someone brings up a cluster where both are disabled and then runs this test?

Should that be supported?

Reply:
If it's not supported, then some comments in the test should explain that, as a kindness towards the poor soul who then presumably gets a test failure and needs to figure out what's causing it.

Reply:
> What happens if someone brings up a cluster where both are disabled and then runs this test?

The intree code will run and the metric will be emitted, the e2e tests will also check for that metric as long as the feature flag setup is also done while calling the e2e tests i.e. setting `CSIMigration=false` in the cluster bootstrap and also `CSIMigration=false` in the e2e tests should work.

Adding some comments about disabling the feature flag in both the cluster and the e2e tests.

Reply:
I don't think it's expected behavior that we don't have any provision metrics with CSI enabled. This sounds like a bug that we need to investigate further.

Reply:
cc @gnufied

Reply:
I created https://github.com/kubernetes/kubernetes/issues/106773 to track the missing metric

Reply:
Let's just remove the check for volume_provision metrics. It is not expected that feature gates are passed into e2e.test and this may cause problems in environments that set feature gates differently from the defaults.

Reply:
ok, I removed the checks